sequencer/src/block2.rs

```diff
@@ -25,51 +25,94 @@ pub struct BlockPayload {
+            tx_table: Vec<u8>,
+            tx_bodies: Vec<u8>, // concatenation of all tx payloads
+            tx_bytes_end: TxTableEntry,
+            tx_table_len: TxTableEntry,
+        }
 
-        // build tx_table, tx_bodies
-        let mut tx_bytes_end = TxTableEntry::zero();
-        let mut tx_table_len = TxTableEntry::zero();
+        // TODO(746) use of HashMap causes nondeterministic ordering for the namespace table.
+        // I think this is ok because a builder can do whatever it wants anyway.
+        let mut namespaces: HashMap<VmId, NamespaceInfo> = HashMap::new();
         for tx in txs.into_iter() {
             let tx_bytes_len: TxTableEntry = tx.payload().len().try_into().ok()?;
-            tx_bytes_end = tx_bytes_end.checked_add(tx_bytes_len)?;
-            tx_table.extend(tx_bytes_end.to_bytes());
-            tx_bodies.extend(tx.payload());
-            tx_table_len = tx_table_len.checked_add(TxTableEntry::one())?;
+
+            let namespace = namespaces.entry(tx.vm()).or_insert(NamespaceInfo {
+                tx_table: Vec::new(),
+                tx_bodies: Vec::new(),
+                tx_bytes_end: TxTableEntry::zero(),
+                tx_table_len: TxTableEntry::zero(),
+            });
+
+            namespace.tx_bytes_end.checked_add_mut(tx_bytes_len)?;
+            namespace.tx_table.extend(namespace.tx_bytes_end.to_bytes());
+            namespace.tx_bodies.extend(tx.payload());
+            namespace
+                .tx_table_len
+                .checked_add_mut(TxTableEntry::one())?;
         }
 
-        // Naively copy all pieces into a flat payload.
-        // We could avoid this by allocating memory in advance,
-        // but that would require knowing the number of txs and their total
-        // byte length in advance, which we can't do without a complete scan
-        // of the `txs` iterator.
+        // first word of namespace table is its length
+        let namespace_table_len = namespaces.len();
+        let mut namespace_table =
+            Vec::from(TxTableEntry::try_from(namespace_table_len).ok()?.to_bytes());
+
+        // fill payload and namespace table
+        // TODO(746) is it worth the trouble to allocate memory in advance?
         let mut payload = Vec::new();
-        payload.extend(tx_table_len.to_bytes());
-        payload.extend(tx_table);
-        payload.extend(tx_bodies);
-        Some(Self {
-            payload,
-            tx_table_len_proof: Default::default(),
-        })
+        let mut namespace_bytes_end = TxTableEntry::zero();
+        for (id, namespace) in namespaces {
+            let prev_len = payload.len();
+            payload.extend(namespace.tx_table_len.to_bytes());
+            payload.extend(namespace.tx_table);
+            payload.extend(namespace.tx_bodies);
+            namespace_table.extend(TxTableEntry::try_from(id).ok()?.to_bytes());
+            namespace_bytes_end.checked_add_mut((payload.len() - prev_len).try_into().ok()?)?;
```

Comment:
I'm adding the length to the previous length using `checked_add_mut`, so if I didn't subtract `prev_len` then I would be adding the total length each time for a quadratic blow-up. It's funny you ask because this was a bug in my earlier code; it's like this because that's what I did to fix the bug. But now that I look at it again, it would be much simpler to not use `checked_add_mut` and then subtract `prev_len`. Instead I could simply *set* it to `payload.len()`. I'll fix that before merging.

Reply:
Oh I see, thanks for explaining. That fix does sound easier to understand